### DISCUSSION TITLE
Infinite number of forest trees (with / without GPU-instancing)

### DISCUSSION THREAD

**Discussion by DerKarlos** [bevyengine/bevy#4496] [Q&A]
My first
Hello all!

I consider to use Bevy for my project (or wgpl only).
The ECS is new to my but no obstacle.
What I read about Bevy looks great. It seems to have all I need,

except GPU-instancing.
Issue #89, how high is it in the roadmap of Bevy?
Could I use wgpl for instances, mixed with Bevy?
(I am rather new to the Rust and crates infrastructure)

A good part of my "game level" consist of a infinite number of forest trees.
What would be the best way without instancing? Is there an example for a forest?
My actual concept is to grid the "level" in tiles
and have one mesh/draw-call for each tree-type(texture) per tile.
A quad-tree may be good to.

Does Bevy offer LoD? Like: Small textures or just a color for far away trees?
Sprites? But I need a correct view from the top to! So I need different sprites, right?

-karlos-

**Comment by tim-blackbird** [comment]
Instancing is possible, but it's boilerplate-heavy at the moment. here is the [example](https://github.com/bevyengine/bevy/blob/main/examples/shader/shader_instancing.rs).

> Does Bevy offer LoD? Like: Small textures or just a color for far away trees?

I don't think Bevy supports mipmaps yet.

> Sprites? But I need a correct view from the top to! So I need different sprites, right?

No built-in billboards yet. You can use quads and manually rotate them to face the player.

**Comment by DerKarlos** [comment]
boilerplate .. yes, but the example works! thank you. I will not understand but try carefully adapt it to my case. That makes me hope.
billboards .. nice. Just change to green will help to. But I assume, the GPU will not do it, bevy neither. I have to. Will make some experiments.

**Comment by DerKarlos** [comment]
My start with bevy is not that easy.
I intend to create my trees simple as 2 or 3 crossing rectangles with textures as it should scale to a really large number of trees.
So I don't really need Normals, but in the code of course.
1) Mesh.duplicate_vertices() and compute_flat_normals() did not work as I am used to it from BabylonJS etc.
So I did it by insert_attribute some values, I found in the bevy code.

2) StandardMaterial { double_sided: true, does not show any difference.
THIS WORKS: cull_mode: None,
? double_sided makes the texture dark, it is pale otherwise

You may see my code here:
https://github.com/DerKarlos/OSMeta/blob/20dff880158448158a24639de9b0331898300917/src/main.rs
And run it here: https://www.osmgo.org/OSMeta/

P.s.: I must admit, some things do work quite well:
                base_color_texture: Some(texture_handle),
                alpha_mode: bevy::pbr::AlphaMode::Mask(0.5),

One more thing:  AlphaMode does NOT work for the shadow 😢
